### PR TITLE
Added code for review (portfolio assignment)

### PR DESCRIPTION
**Pull request by advythr**
I welcome any feedback on my code including readability, improving performance, and modularity.

### CODE REVIEW

File: habitat_code/ramachandran_code_for_review.py [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+
+else:
+    # Mimic web browser
+    user_agent = (
+        'Mozilla/5.0 (X11; Linux x86_64; rv:60.0) '
+        'Gecko/20100101 Firefox/81.0'
+    )
+
+    # Download GEOJSON
+    r = requests.get(url=boundary_url, headers={'User-Agent': user_agent})
```
Comment (peterkobylarz) [inline habitat_code/ramachandran_code_for_review.py:88]: I haven't seen this technique for handling geojson format downloads, nice.

File: habitat_code/ramachandran_code_for_review.py [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+boundary_gdf.to_crs("4326").plot()
+
+# %%
+# Download census tracts
+print(us.states.CO.fips)
+
+colorado_tracts_url = ("https://www2.census.gov/geo/tiger/TIGER2023/"
+           "TRACT/tl_2023_08_tract.zip")
+
+colorado_tracts_path = os.path.join(data_dir, 'colorado_census_tracts.geojson')
```
Comment (peterkobylarz) [inline habitat_code/ramachandran_code_for_review.py:118]: This data is available in .shp format from TIGER I believe. Using .shp would simplify this import process by a few lines. Is there an advantage to using geojson?

File: habitat_code/ramachandran_code_for_review.py [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+
+else:
+    # Authenticate
+    api_key = getpass.getpass('U.S. Census API Key')
+    c = Census(api_key)
+    c
+
+    CO_census = c.acs5.state_county_tract(fields = ('NAME', 'B06011_001E'),
+                                      state_fips = '08',
+                                      county_fips = "*",
```
Comment (peterkobylarz) [inline habitat_code/ramachandran_code_for_review.py:163]: You can supply the County FIPS code here instead of '*' to reduce the data handled. The  us python package lets you do this with the following code: 
```python
county_fips = us.states.lookup(state).fips + us.counties.lookup(county_name).fips
```

File: habitat_code/ramachandran_code_for_review.py [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+all_greenspace_stats_path = os.path.join(data_dir, 'all_greenspace_stats.csv')
+
+if os.path.exists(all_greenspace_stats_path):
+    print("All greenspace stats file exists. Checking "
+          "which tracts have been downloaded and processed...")
+
+    # Extract tracts to be calculated
+    all_tracts = tracts_w_census_gdf['TRACTCE'].astype('int64')
+
+    # Load tracts already processed
```
Comment (peterkobylarz) [inline habitat_code/ramachandran_code_for_review.py:233]: I like this extra caching step to scrape tracts that have already been processed, in case execution is interrupted. Maybe there is some potential to streamline this process with a method?

**Review by peterkobylarz — COMMENTED** [review]
Overall, the code is concise and well commented so I can understand what is going on. There may be some opportunities to streamline the code by using methods. Nice work!

File: habitat_code/ramachandran_code_for_review.py [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+import shapely
+
+import sklearn
+from sklearn.model_selection import train_test_split
+from sklearn.linear_model import LinearRegression
+from sklearn.metrics import mean_squared_error, r2_score
+
+import us
+
+import xarray as xr
```
Comment (edchanonline) [inline habitat_code/ramachandran_code_for_review.py:55]: PEP8 Style - Probably something like

```
import getpass
import io
import os

import cartopy.crs as ccrs
from census import Census
import earthpy as et
import geopandas as gpd
import geoviews as gv
import holoviews as hv
import hvplot
import hvplot.pandas
import hvplot.xarray
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import pystac_client
import requests
import rioxarray as rxr
from rioxarray.merge import merge_arrays
import shapely
import sklearn
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import
```

File: habitat_code/ramachandran_code_for_review.py [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+
+    boundary_gdf = gpd.read_file(boundary_path)
+    print("Data is already downloaded.")
+
+else:
+    # Mimic web browser
+    user_agent = (
+        'Mozilla/5.0 (X11; Linux x86_64; rv:60.0) '
+        'Gecko/20100101 Firefox/81.0'
+    )
```
Comment (edchanonline) [inline habitat_code/ramachandran_code_for_review.py:85]: Nice trick!

File: habitat_code/ramachandran_code_for_review.py (hunk 3 above)
Comment (edchanonline) [reply habitat_code/ramachandran_code_for_review.py:163]: 🔥

File: habitat_code/ramachandran_code_for_review.py [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+
+# %%
+# Calculate and map spatial bias in the model predictions.
+
+analysis_df['pred_fraction_greenspace'] = linear_reg.predict(X)
+analysis_df['err_fraction_greenspace'] = analysis_df['pred_fraction_greenspace'] - analysis_df['fraction_greenspace']
+
+tract_boundaries_gdf = tracts_w_census_gdf[['tract', 'geometry']]
+
+analysis_gdf = pd.merge(tract_boundaries_gdf, analysis_df, on='tract')
```
Comment (edchanonline) [inline habitat_code/ramachandran_code_for_review.py:498]: I remember having better results with GEOID

```
merged_gdf = chi_tracts.merge(census_df, left_on='GEOID', right_on='GEOID', how='inner')
```

File: habitat_code/ramachandran_code_for_review.py [hunk 8]
```diff
@@ -0,0 +1,511 @@
+# %% [markdown]
+# # Boulder, Colorado Urban Greenspace
+# 
+# Boulder, Colorado is a city in the foothills of the Front Range of the Rocky Mountains in the U.S. state of Colorado. It has a semi-arid climate with cold winters and warm, relatively wet summers. The City has been considered a leader in urban forestry being named a Tree City USA by the National Arbor Day Foundation since 1984 [1]. The city maintains a street tree inventory. In 2018, the City of Boulder released the Urban Forest Strategic Plan and in 2023, the City issued a State of the Urban Forest Report [2]. The Strategic Plan outlines a goal to maintain the city's tree canopy at 16% (a no-net-loss goal) given ongoing declines in the tree canopy due to the emerald ash borer.
```
Comment (eculler) [inline habitat_code/ramachandran_code_for_review.py:4]: Nice background -- this really helps put the analysis in context!

File: habitat_code/ramachandran_code_for_review.py (hunk 5 above)
Comment (eculler) [reply habitat_code/ramachandran_code_for_review.py:55]: I don't necessarily mind the way you split this up, @advythr, since it makes it easier to read. I would go ahead and separate out the standard library packages at the top, though, since it helps the reader understand what they might have to install and what is already included.

File: habitat_code/ramachandran_code_for_review.py [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+boundary_gdf
+
+# %%
+boundary_gdf.to_crs("4326").plot()
+
+# %%
+# Download census tracts
+print(us.states.CO.fips)
+
+colorado_tracts_url = ("https://www2.census.gov/geo/tiger/TIGER2023/"
```
Comment (eculler) [inline habitat_code/ramachandran_code_for_review.py:115]: watch the alignment here

File: habitat_code/ramachandran_code_for_review.py [hunk 10]
```diff
@@ -0,0 +1,511 @@
...
+           "TRACT/tl_2023_08_tract.zip")
+
+colorado_tracts_path = os.path.join(data_dir, 'colorado_census_tracts.geojson')
+
+if os.path.exists(colorado_tracts_path):
+    colorado_tracts = gpd.read_file(colorado_tracts_path).to_crs("4326")
+    print("Data is already downloaded.")
+else:
+    gpd.read_file(colorado_tracts_url).to_file(colorado_tracts_path, driver="GeoJSON")
+    colorado_tracts_gdf = gpd.read_file(colorado_tracts_path).to_crs("4326")
```
Comment (eculler) [inline habitat_code/ramachandran_code_for_review.py:125]: This isn't super-repetitive, but it caught my eye because it would be really easy to, say, change the CRS in one spot and not another. In fact, it looks like your code might break if you have to re-download because you have a different variable name (colorado_tracts_gdf vs. colorado_tracts). So watch out for this!

File: habitat_code/ramachandran_code_for_review.py [hunk 11]
```diff
@@ -0,0 +1,511 @@
...
+    census_df = pd.read_csv(census_path)
+    print("Data is already downloaded.")
+
+else:
+    # Authenticate
+    api_key = getpass.getpass('U.S. Census API Key')
+    c = Census(api_key)
+    c
+
+    CO_census = c.acs5.state_county_tract(fields = ('NAME', 'B06011_001E'),
```
Comment (eculler) [inline habitat_code/ramachandran_code_for_review.py:161]: watch the parameter alignment here

File: habitat_code/ramachandran_code_for_review.py [hunk 12]
```diff
@@ -0,0 +1,511 @@
...
+# %%
+# Merge datasets
+
+boulder_city_tracts_gdf = boulder_city_tracts_gdf.loc[:, ['TRACTCE', 'geometry', 'NAMELSAD']]
+
+boulder_city_tracts_gdf['TRACTCE'] = pd.to_numeric(boulder_city_tracts_gdf['TRACTCE'], errors='coerce')
+
+tracts_w_census_gdf = boulder_city_tracts_gdf.merge(census_df, left_on='TRACTCE', right_on='tract')
+
+negative_or_not = (tracts_w_census_gdf['median_income'] >= 0)
```
Comment (eculler) [inline habitat_code/ramachandran_code_for_review.py:194]: I like your variable names, they're super-descriptive! Keep an eye on that line length, though, these are a bit long for my taste, which makes it harder to read.

File: habitat_code/ramachandran_code_for_review.py [hunk 13]
```diff
@@ -0,0 +1,511 @@
...
+# Download and process data for all tracts
+
+ndvi_threshold = 0.12
+
+print("All greenspace stats file does not exist. Running computation.")
+
+all_greenspace_stats = []
+
+# Calculate greenspace fraction for each tract
+for tract, tract_urls in item_url_df.groupby('tract'):
```
Comment (eculler) [inline habitat_code/ramachandran_code_for_review.py:333]: This loop is written very clearly -- I like having the comments and spacing in here. It is, however, a tad long for a single loop, so you might consider splitting it up into a few functions.

**Review by eculler — COMMENTED** [review]
Nice work, @advythr ! This is easy to read, and makes great use of loops to avoid repetition. You might consider some functions as well in future, since that will make it easier to test your work and also easier to debug the loops as needed.
